### PR TITLE
CMake: remove mpi two-pass search

### LINKED ISSUE #5510: Configuring dealii with MPI fails with CMake 3.10.0
I tried to install dealii (Developer version with MPI)  with CMake 3.10.0 (using spack) and the configuring failed.
I get the following output:

```
--   MPI_INCLUDE_DIRS: *** Required variable "MPI_CXX_INCLUDE_PATH" empty ***
--   MPI_USER_INCLUDE_DIRS: *** Required variable "MPI_CXX_INCLUDE_PATH" empty ***
--   MPI_CXX_FLAGS:
--   MPI_LINKER_FLAGS:
-- Could NOT find MPI
-- DEAL_II_WITH_MPI has unmet external dependencies.
CMake Error at cmake/configure/configure_1_mpi.cmake:77 (MESSAGE):

Could not find any suitable mpi library!
```
Using Cmake 3.9.4 there is no problem and installing dealii succeeds.

### PR DESCRIPTION
**Pull request by tamiko**
As discussed in https://github.com/dealii/dealii/issues/5510, we must not
assume anything about the internal workings of CMake's FindMPI.cmake module. In
particular, it is not sufficient to set CMAKE_<LANG>_COMPILER to an
MPI-compiler wrapper in order for FindMPI.cmake to actually detect MPI. So,
remove this logic in our FindMPI.cmake module.

This will force everyone who wishes to use MPI to enable it explicitly via
`-DWITH_MPI=ON` (`-DDEAL_II_WITH_MPI=ON`) on the command line.

TODO:
 - [x] Update documentation
 - [x] add changes entry

Closes #5510

### CODE REVIEW

**Comment by tamiko** [comment]
@bangerth @davydden That seems to be about the only option we have :-/

**Comment by davydden** [comment]
./run-tests

**Comment by tjhei** [comment]
> This will force everyone who wishes to use MPI to enable it explicitly via
> -DWITH_MPI=ON (-DDEAL_II_WITH_MPI=ON) on the command line.

I don't think that is a problem. Looks like a simpler and cleaner solution. :+1: 

By the way, why does ``WITH_MPI=ON`` work without the "DEAL_II_" prefix?

**Comment by bangerth** [comment]
Why is it that we can't test whether MPI is available, and only set it to OFF if it isn't?

**Comment by tamiko** [comment]
> Why is it that we can't test whether MPI is available, and only set it
> to OFF if it isn't?

Well, we (as in Toby, you and me ^^) had a lengthy discussion about this
topic a while ago and the conclusion was to never enable MPI
automatically.

I wouldn't mind reenabling autodetection.

One argument in favor of disabling MPI might have been bthat we had a
lot of users with a broken MPI installation that got picked up
automatically.

**Comment by tamiko** [comment]
> By the way, why does ``WITH_MPI=ON`` work without the "DEAL_II_" prefix?

Because the build system sneakingly prepends DEAL_II_ to all WITH_<...>
variables :-)

**Comment by tjhei** [comment]
> Because the build system sneakingly prepends DEAL_II_ to all WITH_<...>
> variables :-)

Interesting. I think it makes sense to keep advertising only the DEAL_II_* variables.

**Comment by tjhei** [comment]
> One argument in favor of disabling MPI might have been bthat we had a
> lot of users with a broken MPI installation that got picked up
> automatically.

I have always manually specified ``DEAL_II_WITH_MPI=ON`` and I think it is a good idea to keep it that way.

**Comment by bangerth** [comment]
Well, if I agreed with that decision during the discussion you reference, then I guess I should be consistent and stick with my then opinion ;-)

**Comment by bangerth** [comment]
The one test of the `gcc` tester timed out -- not to worry. OK to merge when the second tester is happy.

**Comment by davydden** [comment]
I guess @tamiko still wanted to update some documentation (thus the WIP tag).

**Comment by ChrisTX** [comment]
>One argument in favor of disabling MPI might have been bthat we had a
lot of users with a broken MPI installation that got picked up
automatically.

Just fyi, but CMake 3.10 and newer validate that the values FindMPI return are sufficient to compile a simple MPI test program and otherwise will report `MPI_CXX_FOUND` to be false. It shouldn't be possible anymore for a broken MPI installation to be indicated as "found". CMake 3.10+ also supports determining the MPI version implemented, if you want to use some version variables.

If you want to retain autodetection, you may and depending on the `FOUND` variables should be sufficient on CMake 3.10+. However, keep in mind that this might lead to confusing behavior in that MPI might unintentionally be turned off because something's wrong with the environment.

Lastly, as I understand it, you'd still keep the requirement of `MPI_CXX_LIBRARIES` being non-empty. Keep in mind that this is a valid case if the compiler supports MPI implicitly (e.g. Cray or the compiler wrapper. I would recommend to rely solely on `MPI_<LANG>_FOUND` and consider *all* `MPI_<LANG>_...` variables optional. `MPI_<LANG>_FOUND` can be understood that with the set of `MPI_<LANG>_...` settings will work to provide MPI with `CMAKE_<LANG>_COMPILER`.

**Comment by tamiko** [comment]
@davydden Jup. Will finally get to some deal.II work tomorrow again :-)

**Comment by bangerth** [comment]
What's the status here?

**Comment by tamiko** [comment]
> What's the status here?

Another commit incoming in 10 minutes. After that good to merge.

**Comment by tamiko** [comment]
@bangerth Done. This should finally do the trick (supporting CMake 3.10.x and earlier).

File: cmake/macros/macro_check_mpi_interface.cmake [hunk 1]
```diff
@@ -27,7 +27,7 @@ MACRO(CHECK_MPI_INTERFACE _feature _var)
     SET(_nope FALSE)
 
     FOREACH(_library ${${_feature}_LIBRARIES})
-      IF( _library MATCHES "/libmpi(|_cxx)\\.(a|so)[^/]*$")
+      IF( _library MATCHES "/libmpi\\.(a|so)[^/]*$")
```
Comment (davydden) [inline cmake/macros/macro_check_mpi_interface.cmake]: What about macOS dylib?

Comment (tamiko) [reply cmake/macros/macro_check_mpi_interface.cmake]: @davydden Well, we haven't searched for that particular ending in the past. So I guess the check is useless on Macs. (I would prefer to completely revise the `CHECK_MPI_INTERFACE` macro in a separate PR.)

Comment (davydden) [reply cmake/macros/macro_check_mpi_interface.cmake]: Fine with me

Comment (davydden) [reply cmake/macros/macro_check_mpi_interface.cmake]: Fine with me

Comment (ChrisTX) [reply cmake/macros/macro_check_mpi_interface.cmake]: If you want to generalize this, you can change the regex to use the `CMAKE_<TYPE>_LIBRARY_PREFIX` and  `CMAKE_<TYPE>_LIBRARY_SUFFIX` in the regex, with `<TYPE>` being `SHARED`, `STATIC` or `IMPORT` (e.g. on Windows). For example,  [CMAKE_SHARED_LIBRARY_SUFFIX](https://cmake.org/cmake/help/latest/variable/CMAKE_SHARED_LIBRARY_SUFFIX.html) will be `.so` and [CMAKE_SHARED_LIBRARY_PREFIX](https://cmake.org/cmake/help/latest/variable/CMAKE_SHARED_LIBRARY_PREFIX.html) will be `lib` on Linux. That should yield the most flexible solution.

Be aware though that not all MPIs are incompatible, e.g. there's the [MPICH ABI initiative](http://www.mpich.org/abi/).

**Pull request merged**